src/backend/commands/copy.c

```diff
@@ -722,34 +723,92 @@ CopyGetData(CopyState cstate, void *databuf, int datasize)
 	switch (cstate->copy_dest)
 	{
 		case COPY_FILE:
-			bytesread = fread(databuf, 1, datasize, cstate->copy_file);
-			if (feof(cstate->copy_file))
-				cstate->fe_eof = true;
-			if (ferror(cstate->copy_file))
 			{
-				if (cstate->is_program)
+				struct pollfd pollfd;
+				int			rc;
+
+				/* Initialize pollfd */
+				pollfd.fd = fileno(cstate->copy_file);
+				pollfd.events = POLLIN;
+
+				/*
+				 * Wait until we get data, even if interrupted by a signal,
+				 * while also checking for interrupts
+				 */
+				do
 				{
-					int olderrno = errno;
+					CHECK_FOR_INTERRUPTS();
 
-					close_program_pipes(cstate, true);
+					/* Wait until data arrives or 500 ms passes */
+#define COPY_POLL_TIMEOUT 500
+					pollfd.revents = 0;
+					rc = poll(&pollfd, 1, COPY_POLL_TIMEOUT);
```

Comment:
consider using WaitLatchOrSocket instead poll

Reply:
Looked into it, `WaitLatchOrSocket` immediately calls `ereport` on error, so no way to do `close_program_pipes`. Should I really replace that?

Reply:
Maybe PG_CATCH will help?

Reply:
Done